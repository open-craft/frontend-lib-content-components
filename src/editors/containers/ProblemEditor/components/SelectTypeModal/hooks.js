import {
  useState,
} from 'react';
import { StrictDict } from '../../../../utils';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
