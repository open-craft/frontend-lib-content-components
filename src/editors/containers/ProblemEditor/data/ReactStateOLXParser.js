import _ from 'lodash-es';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import { ProblemTypeKeys } from '../../../data/constants/problem';

export class ReactStateOLXParser {
  constructor(problemState) {
    const parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      // preserveOrder: true
    };
    const builderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
    };
    this.parser = new XMLParser(parserOptions);
    this.builder = new XMLBuilder(builderOptions);
    this.problemState = problemState.problem;
  }

  addHints() {
    const hintsArray = [];
    const hints = _.get(this.problemState, 'settings.hints', []);
    hints.forEach(element => {
      hintsArray.push({
        '#text': element.value,
      });
    });
    const demandhint = {
      demandhint: {
        hint: hintsArray,
      },
    };
    return demandhint;
  }

  addMultiSelectAnswers(option) {
    const choice = [];
    let compoundhint = [];
    let widget = {};
    const { answers } = this.problemState;
    answers.forEach((answer) => {
      const feedback = [];
      let singleAnswer = {};
      if (_.has(answer, 'selectedFeedback') && _.get(answer, 'selectedFeedback', '').trim() !== '') {
        feedback.push({
          '#text': _.get(answer, 'selectedFeedback'),
          '@_selected': true,
        });
      }
      if (_.has(answer, 'unselectedFeedback') && _.get(answer, 'unselectedFeedback', '').trim() !== '') {
        feedback.push({
          '#text': _.get(answer, 'unselectedFeedback'),
          '@_selected': false,
        });
      }
      if (_.has(answer, 'feedback') && _.get(answer, 'feedback', '').trim() !== '') {
        feedback.push({
          '#text': _.get(answer, 'feedback'),
        });
      }
      if (feedback.length) {
        singleAnswer[`${option}hint`] = feedback;
      }
      singleAnswer = {
        '#text': answer.title,
        '@_correct': answer.correct,
        ...singleAnswer,
      };
      choice.push(singleAnswer);
    });
    widget = { [option]: choice };
    if (_.has(this.problemState, 'groupFeedbackList')) {
      compoundhint = this.addGroupFeedbackList();
      widget = {
        ...widget,
        compoundhint,
      };
    }
    return widget;
  }

  addGroupFeedbackList() {
    const compoundhint = [];
    const { groupFeedbackList } = this.problemState;
    groupFeedbackList.forEach((element) => {
      compoundhint.push({
        '#text': element.feedback,
        '@_value': element.answers.join(' '),
      });
    });
    return compoundhint;
  }

  addQuestion() {
    const { question } = this.problemState;
    const questionObject = this.parser.parse(question);
    return questionObject;
  }

  buildMultiSelectProblem(problemType, widget, option) {
    const question = this.addQuestion();
    const widgetObject = this.addMultiSelectAnswers(option);
    const demandhint = this.addHints();
    const problemObject = {
      problem: {
        [problemType]: {
          ...question,
          [widget]: widgetObject,
        },
        ...demandhint,
      },
    };
    return this.builder.build(problemObject);
  }

  buildTextInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildTextInputAnswersFeedback();
    const problemObject = {
      problem: {
        [ProblemTypeKeys.TEXTINPUT]: {
          ...question,
          ...answerObject,
        },
        ...demandhint,
      },
    };
    return this.builder.build(problemObject);
  }

  buildTextInputAnswersFeedback() {
    const { answers } = this.problemState;
    let answerObject = {};
    const additionAnswers = [];
    const wrongAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (answer.correct && firstCorrectAnswerParsed) {
        additionAnswers.push({
          '@_answer': answer.title,
          ...correcthint,
        });
      } else if (answer.correct && !firstCorrectAnswerParsed) {
        firstCorrectAnswerParsed = true;
        answerObject = {
          '@_answer': answer.title,
          ...correcthint,
        };
      } else if (!answer.correct) {
        wrongAnswers.push({
          '@_answer': answer.title,
          '#text': answer.feedback,
        });
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionAnswers,
      stringequalhint: wrongAnswers,
      '@_type': _.get(this.problemState, 'additionalAttributes.type', 'ci'),
      textline: {
        '@_size': _.get(this.problemState, 'additionalAttributes.textline.size', 20),
      },
    };
    return answerObject;
  }

  buildNumericInput() {
    const question = this.addQuestion();
    const demandhint = this.addHints();
    const answerObject = this.buildNumericalResponse();
    const problemObject = {
      problem: {
        ...question,
        [ProblemTypeKeys.NUMERIC]: answerObject,
        ...demandhint,
      },
    };
    return this.builder.build(problemObject);
  }

  buildNumericalResponse() {
    const { answers } = this.problemState;
    let answerObject = {};
    const additionalAnswers = [];
    let firstCorrectAnswerParsed = false;
    /*
    TODO: Need to figure out how to add multiple numericalresponse,
    the parser right now converts all the other right answers into
    additional answers.
    */
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (answer.correct && !firstCorrectAnswerParsed) {
        firstCorrectAnswerParsed = true;
        let responseParam = {};
        if (_.has(answer, 'tolerance')) {
          responseParam = {
            responseparam: {
              '@_type': 'tolerance',
              '@_default': _.get(answer, 'tolerance', 0),
            },
          };
        }
        answerObject = {
          '@_answer': answer.title,
          ...responseParam,
          ...correcthint,
        };
      } else if (answer.correct && firstCorrectAnswerParsed) {
        additionalAnswers.push({
          '@_answer': answer.title,
          ...correcthint,
        });
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionalAnswers,
      formulaequationinput: {
        '#text': '',
      },
    };
    return answerObject;
  }

  getAnswerHints(elementObject) {
    const feedback = elementObject?.feedback;
    let correcthint = {};
    if (feedback !== undefined && feedback !== '') {
      correcthint = {
        correcthint: {
          '#text': feedback,
        },
      };
    }
    return correcthint;
  }

  buildOLX() {
    const { problemType } = this.problemState;
    let problemString = '';
    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.MULTISELECT, 'checkboxgroup', 'choice');
        break;
      case ProblemTypeKeys.DROPDOWN:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.DROPDOWN, 'optioninput', 'option');
        break;
      case ProblemTypeKeys.SINGLESELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.SINGLESELECT, 'choicegroup', 'choice');
        break;
      case ProblemTypeKeys.TEXTINPUT:
        problemString = this.buildTextInput();
        break;
      case ProblemTypeKeys.NUMERIC:
        problemString = this.buildNumericInput();
        break;
      default:
        break;
    }
    return problemString;
  }
}
