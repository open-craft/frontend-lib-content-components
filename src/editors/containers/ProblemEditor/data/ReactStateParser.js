import _ from "lodash";
import { ProblemTypeKeys } from "../../../data/constants/problem";
import { indexToLetterMap } from "./MarkDownParser";


export class ReactStateParser{
    constructor(problemState){
        this.problemState = problemState;
    }

    parseQuestion() {
        let question = this.problemState.question;
        let paresedQuestion = this.splitLines(question);
        let parsedString = '';
        for (const line in paresedQuestion) {
            let replacedString = paresedQuestion[line].replace("<p>", ">>").replace("</p>", "<<");
            parsedString += `${replacedString}\n`
        }
        return parsedString;
    }

    parseHints() {
        let hintString = '';
        let hitnsList = this.problemState.hints;
        for (const hint in hitnsList){
            hintString += `||${hitnsList[hint].value}||\n`;
        }
        return hintString
    }

    parseDropdown(){
        let answerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerOptionString = '';
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            if (answerObject.correct) {
                answerOptionString += ` (${answerContent})`
            } else {
                answerOptionString += ` ${answerContent}`
            }
            if (answerFeedback !== undefined && answerFeedback.trim() !== ''){
                answerOptionString += ` {{${answerFeedback}}}`
            }
            answerString += `${answerOptionString}\n`
        }
        answerString = `[[\n${answerString}]]\n`;
        return answerString;
    }

    parseCheckBox(){
        let anwerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerSelectedFeedback = answerObject?.selectedFeedback;
            let answerUnselectedFeedback = answerObject?.unselectedFeedback;
            let isSelectedFeedbackPresent = (answerSelectedFeedback.trim() !== '' && answerSelectedFeedback !== undefined);
            let isUnselectedFeedbackPresent = (answerUnselectedFeedback.trim() !== '' && answerUnselectedFeedback !== undefined);
            if (answerObject.correct) {
                anwerString += `[x] ${answerContent}`
            } else {
                anwerString += `[ ] ${answerContent}`
            }
            if (isSelectedFeedbackPresent && isUnselectedFeedbackPresent){
                anwerString += `{{selected: ${answerSelectedFeedback}},{unselected: ${answerUnselectedFeedback}}}`
            } else if (isSelectedFeedbackPresent) {
                anwerString += `{{selected: ${answerSelectedFeedback}}}`
            } else if (isUnselectedFeedbackPresent) {
                anwerString += `{{unselected: ${answerUnselectedFeedback}}}`
            }
            anwerString += "\n";
        }
        return anwerString;
    }


    parseRadioButton(){
        let answerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerOptionString = '';
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            if (answerObject.correct) {
                answerOptionString += `(x) ${answerContent}`;
            } else {
                answerOptionString += `( ) ${answerContent}`;
            }
            if (answerFeedback.trim() !== '' && answerFeedback !== undefined){
                answerOptionString += ` {{${answerFeedback}}}`;
            }
            answerString += `${answerOptionString}\n`;
        }
        return answerString;
    }

    parseInputAnswers(){
        let answerString = '';
        let answers = this.problemState.answers;
        let correctAnswer = false;
        for (const answer in answers){
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            let answerOptionString = ''
            if (answerObject.correct && !correctAnswer) {
                answerOptionString += `=${answerContent}`;
                correctAnswer = true;
            } else if (answerObject.correct && correctAnswer) {
                answerOptionString += `or=${answerContent}`;
            } else if (!answerObject.correct){
                answerOptionString += `not=${answerContent}`;
            }
            if (answerFeedback.trim() !== '' && answerFeedback !== undefined){
                answerOptionString += ` {{${answerFeedback}}}`;
            }
            answerString += `${answerOptionString}\n`;
        }
        return answerString;
    }

    parseGroupFeedback() {
        let feedbackString = "";
        const groupFeedbackArray = this.problemState.groupFeedbackList;
        if (!_.isEmpty(groupFeedbackArray)){
            for (const groupFeeback in groupFeedbackArray) {
                let answers = groupFeedbackArray[groupFeeback].answers;
                let options = `(( ${answers.join(" ")} ))`;
                feedbackString += `{{ ${options} ${groupFeedbackArray[groupFeeback].feedback} }}\n`;
            }
        }
        return feedbackString
    }

    getMarkdown() {
        let answers = ''
        switch (this.problemState.problemType) {
            case ProblemTypeKeys.DROPDOWN:
                answers = this.parseDropdown();
                break;
            case ProblemTypeKeys.MULTISELECT:
                answers = this.parseCheckBox();
                break;
            case ProblemTypeKeys.SINGLESELECT:
                answers = this.parseRadioButton();
                break;
            case ProblemTypeKeys.NUMERIC:
            case ProblemTypeKeys.TEXTINPUT:
                answers = this.parseInputAnswers();
                break;
            default:
                break;
        }
        let markdown = `${this.parseQuestion()}${answers}${this.parseHints()}${this.parseGroupFeedback()}`;
        return markdown;
    }

    splitLines(inputString){
        return inputString.trim().split("\n");
    }
}
