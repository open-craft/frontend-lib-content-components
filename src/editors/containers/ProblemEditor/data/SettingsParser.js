import { isEmpty, get, isNil } from 'lodash-es';

import { ShowAnswerTypes, RandomizationType } from "../../../data/constants/problem";

export const popuplateItem = (parentObject, itemName, statekey, metadata) => {
    let item = get(metadata, itemName, null);
    if (!isNil(item)) {
        parentObject = { ...parentObject, [statekey]: item };
    }
    return parentObject;
}

export const parseScoringSettings = (metadata) => {
    let scoring = {};

    let attempts = popuplateItem({}, 'max_attempts', 'number', metadata);
    if (!isEmpty(attempts)){
        let unlimited = true;
        if (attempts.number > 0) {
            unlimited = false;
        }
        attempts = { ...attempts, unlimited: unlimited, };
        scoring = { ...scoring, attempts };
    }

    scoring = popuplateItem(scoring, 'weight', 'weight', metadata);

    return scoring;
}

export const parseShowAnswer = (metadata) => {
    let showAnswer = {};

    let showAnswerType = get(metadata, 'showanswer', {});
    if (!isNil(showAnswerType) && showAnswerType in ShowAnswerTypes){
        showAnswer = { ...showAnswer, ['on']: showAnswerType };
    }

    showAnswer = popuplateItem(showAnswer, 'attempts_before_showanswer_button', 'afterAttempts', metadata)

    return showAnswer;
}

export const parseSettings = (metadata) => {
    let settings = {};
    
    if (isNil(metadata) || isEmpty(metadata)){
        return settings;
    }

    settings = popuplateItem(settings, 'matlab_api_key', 'matLabApiKey', metadata);

    let scoring = parseScoringSettings(metadata);
    if (!isEmpty(scoring)){
        settings = { ...settings, scoring };
    }

    let randomization = get(metadata, 'rerandomize', null)
    if (!isNil(randomization) && randomization in RandomizationType){
        settings = { ...settings, randomization };
    }

    let showAnswer = parseShowAnswer(metadata);
    if (!isEmpty(showAnswer)){
        settings = { ...settings, showAnswer }
    }

    settings = popuplateItem(settings, 'show_reset_button', 'showResetButton', metadata);
    settings = popuplateItem(settings, 'submission_wait_seconds', 'timeBetween', metadata);

    return settings;
}
  