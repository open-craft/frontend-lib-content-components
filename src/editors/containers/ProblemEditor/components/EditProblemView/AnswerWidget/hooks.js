import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback,
};
