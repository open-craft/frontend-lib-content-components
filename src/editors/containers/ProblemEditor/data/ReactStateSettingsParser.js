import { popuplateItem } from "./SettingsParser";

export class ReactStateSettingsParser {
  constructor(problemState) {
    this.problemState = problemState;
  }

  getSettings() {
    let settings = {}
    const stateSettings = this.problemState.settings

    settings = popuplateItem(settings, 'matLabApiKey', 'matlab_api_key', stateSettings)
    settings = popuplateItem(settings, 'number', 'max_attempts', stateSettings.scoring.attempts);
    settings = popuplateItem(settings, 'weight', 'weight', stateSettings.scoring);
    settings = popuplateItem(settings, 'randomization', 'rerandomize', stateSettings);
    settings = popuplateItem(settings, 'on', 'showanswer', stateSettings.showAnswer);
    settings = popuplateItem(settings, 'afterAttempts', 'attempts_before_showanswer_button', stateSettings.showAnswer);
    settings = popuplateItem(settings, 'showResetButton', 'show_reset_button', stateSettings);
    settings = popuplateItem(settings, 'timeBetween', 'submission_wait_seconds', stateSettings);

    return settings;
  }

}
