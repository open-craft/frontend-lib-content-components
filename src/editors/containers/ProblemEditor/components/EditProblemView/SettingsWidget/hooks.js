import { useState, useEffect } from 'react';

import * as module from './hooks';
import messages from './messages';
import { actions } from '../../../../../data/redux';
import { isEmpty } from 'lodash-es';

export const state = {
    showAdvanced: (val) => useState(val),
    cardCollapsed: (val) => useState(val),
    summary: (val) => useState(val),
};

export const showAdvancedSettingsCards = () => {
    const [isAdvancedCardsVisible, setIsAdvancedCardsVisible] = module.state.showAdvanced(false);
    return {
        isAdvancedCardsVisible,
        showAdvancedCards: () => setIsAdvancedCardsVisible(true),
    };
}

export const showFullCard = () => {
    const [isCardCollapsed, setIsCardCollapsed] = module.state.cardCollapsed(false);
    return {
        isCardCollapsed,
        toggleCardCollapse: () => setIsCardCollapsed(!isCardCollapsed),
    };
}

export const hintsCardHooks = (hints, dispatch) => {
    const [summary, setSummary] = module.state.summary({ message: "", values: {} })

    useEffect(() => {
        const hintsNumber = hints.length;
        if (hintsNumber == 0) {
            setSummary({ message: messages.noHintSummary, values: {} });
        } else {
            setSummary({ message: messages.hintSummary, values: { hint: hints[0].value, count: (hintsNumber - 1) } });
        }
    }, [hints]);

    const handleAdd = () => {
        let newId = Math.max(...hints.map(hint => hint.id)) + 1
        const hint = { id: newId, value: "" }
        hints = [...hints, hint]
        dispatch(actions.problem.updateSettings({ hints }));
    }


    return {
        summary,
        handleAdd,
    }
}

export const hintsRowHooks = (id, hints, dispatch) => {

    const handleChange = (event) => {
        const value = event.target.value
        hints = hints.map(hint => {
            if (hint.id === id) {
                return { ...hint, value };
            }
            return hint;
        });
        dispatch(actions.problem.updateSettings({ hints }));
    }

    const handleDelete = () => {
        hints = hints.filter((hint) => (hint.id != id));
        dispatch(actions.problem.updateSettings({ hints }));
    }

    return {
        handleChange,
        handleDelete,
    }
}

export const matlabCardHooks = (matLabApiKey, dispatch) => {
    const [summary, setSummary] = module.state.summary({ message: "", values: {}, intl: false })

    useEffect(() => {
        if (isEmpty(matLabApiKey)) {
            setSummary({ message: messages.matlabNoKeySummary, values: {}, intl: true });
        } else {
            setSummary({ message: matLabApiKey, values: {}, intl: false });
        }
    }, [matLabApiKey])

    const handleChange = (event) => {
        dispatch(actions.problem.updateSettings({ matLabApiKey: event.target.value }));
    }

    return {
        summary,
        handleChange,
    }
}

export const randomizationCardHooks = (dispatch) => ({
    handleChange: (event) => {
        dispatch(actions.problem.updateSettings({ randomization: event.target.value }));
    }
})

export const resetCardHooks = (dispatch) => {

    const setReset = (value) => {
        dispatch(actions.problem.updateSettings({ showResetButton: value }));
    }

    return {
        setResetTrue: () => setReset(true),
        setResetFalse: () => setReset(false),
    }
}

export const scoringCardHooks = (scoring, dispatch) => {

    const handleMaxAttemptChange = (event) => {
        let unlimitedAttempts = true;
        const attemptNumber = parseInt(event.target.value);
        if (attemptNumber > 0) {
            unlimitedAttempts = false;
        }
        dispatch(actions.problem.updateSettings({ scoring: { ...scoring, attempts: { number: attemptNumber, unlimited: unlimitedAttempts } } }));
    }

    const handleWeightChange = (event) => {
        dispatch(actions.problem.updateSettings({ scoring: { ...scoring, weight: parseFloat(event.target.value) } }));
    }

    return {
        handleMaxAttemptChange,
        handleWeightChange,
    }
}

export const showAnswerCardHooks = (showAnswer, dispatch) => {

    const handleShowAnswerChange = (event) => {
        dispatch(actions.problem.updateSettings({ showAnswer: { ...showAnswer, on: event.target.value } }));
    }

    const handleAttemptsChange = (event) => {
        dispatch(actions.problem.updateSettings({ showAnswer: { ...showAnswer, afterAttempts: parseInt(event.target.value) } }));
    }

    return {
        handleShowAnswerChange,
        handleAttemptsChange,
    }
}

export const timerCardHooks = (dispatch) => ({

    handleChange: (event) => {
        dispatch(actions.problem.updateSettings({ timeBetween: event.target.value }));
    }
})

export const typeRowHooks = (typeKey, dispatch) => {

    const onClick = () => {
        dispatch(actions.problem.updateField({ problemType: typeKey }));
    }

    return {
        onClick,
    }

}