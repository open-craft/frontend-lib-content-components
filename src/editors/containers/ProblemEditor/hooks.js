import { useRef, useCallback, useState, useEffect } from 'react';
import { StrictDict } from '../../utils';
import * as module from './hooks';

export const state = StrictDict({
    refReady: (val) => useState(val),
  });

export const problemEditorConfig = ({
    setEditorRef,
    editorRef,
    question,
    updateQuestion
  }) => ({
    onInit: (evt, editor) => {
      setEditorRef(editor);
    },
    initialValue: question ? question : '',
    onFocusOut: () =>{
      let content = editorRef.current.getContent({format: 'text'});
      updateQuestion(content);
    },
  });

  export const prepareEditorRef = () => {
    const editorRef = useRef(null);
    const setEditorRef = useCallback((ref) => {
      editorRef.current = ref;
    }, []);
    const [refReady, setRefReady] = module.state.refReady(false);
    useEffect(() => setRefReady(true), []);
    return { editorRef, refReady, setEditorRef };
  };
