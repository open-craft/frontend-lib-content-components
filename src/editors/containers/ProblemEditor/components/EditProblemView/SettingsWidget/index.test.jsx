import React from 'react';
import { shallow } from 'enzyme';
import { showAdvancedSettingsCards } from './hooks';
import { SettingsWidget, mapDispatchToProps } from '.';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';
import { actions } from '../../../../../data/redux';

jest.mock('./hooks', () => ({
  showAdvancedSettingsCards: jest.fn(),
}));

describe('SettingsWidget', () => {
  const props = {
    problemType: ProblemTypeKeys.TEXTINPUT,
    settings: {},
  };

  describe('behavior', () => {
    it(' calls showAdvancedSettingsCards when initialized', () => {
      const showAdvancedSettingsCardsProps = {
        isAdvancedCardsVisible: false,
        setResetTrue: jest.fn().mockName('showAdvancedSettingsCards.setResetTrue'),
      };
      showAdvancedSettingsCards.mockReturnValue(showAdvancedSettingsCardsProps);
      shallow(<SettingsWidget {...props} />);
      expect(showAdvancedSettingsCards).toHaveBeenCalledWith();
    });
  });

  describe('snapshot', () => {
    test('snapshot: renders Settings widget page', () => {
      const showAdvancedSettingsCardsProps = {
        isAdvancedCardsVisible: false,
        setResetTrue: jest.fn().mockName('showAdvancedSettingsCards.setResetTrue'),
      };
      showAdvancedSettingsCards.mockReturnValue(showAdvancedSettingsCardsProps);
      expect(shallow(<SettingsWidget {...props} />)).toMatchSnapshot();
    });
    test('snapshot: renders Settings widget page advanced settings visible', () => {
      const showAdvancedSettingsCardsProps = {
        isAdvancedCardsVisible: true,
        setResetTrue: jest.fn().mockName('showAdvancedSettingsCards.setResetTrue'),
      };
      showAdvancedSettingsCards.mockReturnValue(showAdvancedSettingsCardsProps);
      expect(shallow(<SettingsWidget {...props} />)).toMatchSnapshot();
    });
  });

  describe('mapDispatchToProps', () => {
    test('updateSettings from actions.problem.updateSettings', () => {
      expect(mapDispatchToProps.updateSettings).toEqual(actions.problem.updateSettings);
    });
  });

  describe('mapDispatchToProps', () => {
    test('updateField from actions.problem.updateField', () => {
      expect(mapDispatchToProps.updateField).toEqual(actions.problem.updateField);
    });
  });
});
