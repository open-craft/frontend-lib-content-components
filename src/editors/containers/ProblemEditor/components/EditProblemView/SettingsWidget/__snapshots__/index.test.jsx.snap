// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsWidget snapshot snapshot: renders Settings widget page 1`] = `
<div>
  <div>
    <h3>
      <FormattedMessage
        defaultMessage="Settings"
        description="Settings Title"
        id="authoring.problemeditor.settings.settingsWidgetTitle"
      />
    </h3>
    <Container>
      <Row>
        <Col>
          <Row
            className="my-2"
          >
            <InjectIntl(ShimmedIntlComponent)
              problemType="TEXTINPUT"
            />
          </Row>
          <Row
            className="my-2"
          >
            <InjectIntl(ShimmedIntlComponent) />
          </Row>
          <Row
            className="mt-2"
          >
            <InjectIntl(ShimmedIntlComponent) />
          </Row>
          <Row>
            <Collapsible.Advanced
              open={true}
            >
              <Collapsible.Body
                className="collapsible-body"
              >
                <Button
                  className="my-3 ml-2"
                  size="inline"
                  variant="link"
                >
                  <FormattedMessage
                    defaultMessage="Show advanced settings"
                    description="Button text to show advanced settings"
                    id="authoring.problemeditor.settings.showAdvancedButton"
                  />
                </Button>
              </Collapsible.Body>
            </Collapsible.Advanced>
          </Row>
          <Collapsible.Advanced
            open={false}
          >
            <Collapsible.Body
              className="collapsible-body"
            >
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
            </Collapsible.Body>
          </Collapsible.Advanced>
        </Col>
      </Row>
    </Container>
  </div>
</div>
`;

exports[`SettingsWidget snapshot snapshot: renders Settings widget page advanced settings visible 1`] = `
<div>
  <div>
    <h3>
      <FormattedMessage
        defaultMessage="Settings"
        description="Settings Title"
        id="authoring.problemeditor.settings.settingsWidgetTitle"
      />
    </h3>
    <Container>
      <Row>
        <Col>
          <Row
            className="my-2"
          >
            <InjectIntl(ShimmedIntlComponent)
              problemType="TEXTINPUT"
            />
          </Row>
          <Row
            className="my-2"
          >
            <InjectIntl(ShimmedIntlComponent) />
          </Row>
          <Row
            className="mt-2"
          >
            <InjectIntl(ShimmedIntlComponent) />
          </Row>
          <Row>
            <Collapsible.Advanced
              open={false}
            >
              <Collapsible.Body
                className="collapsible-body"
              >
                <Button
                  className="my-3 ml-2"
                  size="inline"
                  variant="link"
                >
                  <FormattedMessage
                    defaultMessage="Show advanced settings"
                    description="Button text to show advanced settings"
                    id="authoring.problemeditor.settings.showAdvancedButton"
                  />
                </Button>
              </Collapsible.Body>
            </Collapsible.Advanced>
          </Row>
          <Collapsible.Advanced
            open={true}
          >
            <Collapsible.Body
              className="collapsible-body"
            >
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
              <Row
                className="my-2"
              >
                <InjectIntl(ShimmedIntlComponent) />
              </Row>
            </Collapsible.Body>
          </Collapsible.Advanced>
        </Col>
      </Row>
    </Container>
  </div>
</div>
`;
